src/pyodide-py/pyodide/webloop.py

```diff
@@ -0,0 +1,429 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+# flake8: noqa
+
+"""
+
+A custom event loop for running asyncio in Pyodide
+
+Version: 0.1.2
+
+Author: Wei OUYANG (@oeway)
+        Adapted from the EventSimulator made by @damonjw
+        https://gist.github.com/damonjw/35aac361ca5d313ee9bf79e00261f4ea
+
+Version History:
+        * support contextvars
+        * add event loop policy
+        * throw error when the loop is already running
+        * support Python 3.8
+        * fix run_until_complete
+"""
+
+import heapq
+import asyncio
+import time
+import traceback
+import contextvars
+
+import js
+from typing import Dict, Tuple, Optional, Awaitable, Callable
+
+
+class WebLoop(asyncio.AbstractEventLoop):
+    """A custom event loop for running asyncio in Pyodide
+
+    It works by utilizing the browser event loop via the setTimeout function
+    """
+
+    def __init__(self, debug: Optional[bool] = False, interval: Optional[int] = 10):
+        """
+        Instatiate the web loop
+        """
+        self._running = False
+        self._immediate = []
```

Comment:
Maybe use a `collections.deque` since the only operations on it are append and pop. That would also avoid taking slices to pop the first element..